### ISSUE TITLE
[RFE] add "exclude" option for install

### ISSUE THREAD

**Issue by remicollet** [composer/composer#4438]
"composer install" use the github generated archive to get the component content.

Lot of projects start to consider that providing the "test suite" is not needed :(
So they exclude various file using the .gitattributes feature.

This seems to be a terrible practice, the test suite have to be in the project archive, it is part of the project, have value for users, and can even be considered as documentation.

Using .gitattributes, used to solved a composer need, should be manage in compopser.

Proposal : extends the schema to allow

     "exclude": [
        "phpunit.xml",
        "tests/"
    ],

This will make .gitattributes bas usage deprecated, and will respect good practique, and will make everyone life easier :)

**Comment by Ocramius** [comment]
The test suite, as well as any other build-related artifacts, is indeed not needed for packages.
The packages are also (in most cases) not designed to be tested from an installation dir in `vendor`.

The test suite in vendor is confusing users, and doing the complete opposite of providing value.

Exactly like in other projects/language, I see zero issues in differentiating between source and distribution.

**Comment by remicollet** [comment]
> I see zero issues in differentiating between source and distribution.

@Ocramius please clarify what you mean

**Comment by Ocramius** [comment]
Also clarifying: you are posting this (again) because the build tools for RPM/DEB are using zip downloads from GH. Just use git directly (as should be), so you can actually export signed tags information as well instead?

It seems like we're trying to fix a symptom (rather than the cause) here.

**Comment by Ocramius** [comment]
@remicollet

Source: all of the project
Dist: just what is needed at runtime

**Comment by bochecha** [comment]
> Exactly like in other projects/language, I see zero issues in differentiating between source and distribution.

Sure, that's why every other projects/languages provide source tarball which contain everything (including tests), and some install procedure (e.g `make install` or `python setup.py install` or ...) that won't install the unit tests.

**Comment by Ocramius** [comment]
Right, and that's `--prefer-source` here: works brilliantly.

**Comment by remicollet** [comment]
> Also clarifying: you are posting this (again) because the build tools for RPM/DEB are using zip downloads from GH

Not again, 1st time here (after trying to convince various upstream of their bad practice), as I think this is really something which need to be fixed globally, so in composer.

And RPM build tools can use everything, including a git snapshot, this is not a RPM issue.

> Source: all of the project
> Dist: just what is needed at runtime

I definitively agree:
* sources = **github tarball** = all of the project
* dist = **what is installed by composer** = just what is needed at runtime

**Comment by Ocramius** [comment]
So what is the issue with `.gitattributes` specifically? You are suggesting an alternate config for excludes, but that doesn't work with github's infra, so it actually creates a new problem (hosting those partial packages).

**Comment by remicollet** [comment]
@Ocramius Have you read the initial description  ? I suggest to add this is composer.json, as this is a composer feature, and then can be clean from .gitattributes (in all projects)

**Comment by Ocramius** [comment]
I still think you are just re-introducing an already solved problem (bandwidth+space) to solve a non-existing problem (source packages, available via git anyway).

**Comment by jakoch** [comment]
This "exclude" feature has been requested several times before, ref. #1750.
I consider "excluding" being a build task. Its one of the steps, when going from "Source" to "Dist",
where "Source" is all of the project (a developer release) and "Dist", just what is needed at runtime (a production release).

How can Composer help to get from Source to Dist? Well, there is the ["archive" command](https://getcomposer.org/doc/04-schema.md#archive) 
and (in regard to the feature request) it already takes already takes excludes into account.
Would that be a fit for your use-case, Remi?

If you want to save bandwidth and shrink the package size (before Composer download), build properly. If you want to just clean things up (after Composer download), integrate a cleanup step into the application installation or bootstrap process.

**Comment by alcohol** [comment]
We've discussed this before. Stop trying to enforce your personal opinion on other peoples projects.

**Comment by remicollet** [comment]
> We've discussed this before. 

Again no. Yes I have ask some projects to provide the test suite. This feature is for "composer" itself to manage this properly.

Please look around you
* **perl** libraries available on http://www.cpan.org/ provide their test suite (and cpan command is able to manage them and not install them)
* **python** libraries available on https://pypi.python.org/pypi provide their test suite (and pip command is able to manage them)
* **ruby** libraries available on https://rubygems.org/ provide their test suite (and gem command is able to manage them)
* **nodejs** libraries available on https://www.npmjs.com/ provide their test suite (and npm command is able to manage them)
* **php** libraries available http://pear.php.net/ provide their test suite (and yes the 20 years old pear command is able to manage them)
* etc

So for "ALL" languages, sources contains everything,  and tooling manage installation of what is needed using the metadata.

For composer libraries, the metadata are in composer.json, and the tooling is the "composer" command.

To summarize, this proposal is to align "composer" with all other languages tooling.

Sorry, but I definitively don't understand why a PHP library should be managed differently.

**Comment by Ocramius** [comment]
Could you clarify exactly on the advantages of having the test suite in a dist package?

Since test suites are hardly working in the same way across all packages, I don't see a way for a consumer of a dist package to run the test suite (even if it's there).

Besides "aligning to other languages", I don't see anything useful coming from this.

If there was a "post-install smoke-test" system, then fine, but that's not going to happen anyway due to security implications.

**Comment by pypingou** [comment]
> Could you clarify exactly on the advantages of having the test suite in a dist package?

People can run them?

**Comment by Ocramius** [comment]
As I said above, people can't run them anyway, as the test suite won't work in isolation most of the times anyway.

**Comment by pypingou** [comment]
> As I said above, people can't run them anyway, as the test suite won't work in isolation most of the times anyway.

I don't see how this is composer's problem it's hardly something to investigate/solve between upstream and its users

**Comment by Ocramius** [comment]
@pypingou most test suites include integration tests as well.

Anyway, running test suites is usually a dev-time task, which requires dev dependencies to be installed as well, and therefore it simply wouldn't work within a dist package inside `vendor`.

That said, I'm unsubscribing from this discussion, as it seems to go nowhere and provide no good reasons to go there (this place called "nowhere") anyway.

**Comment by alcohol** [comment]
My answer remains no. Using .gitattributes for this is a perfectly reasonable solution.

**Comment by alexpott** [comment]
So one reason why I think composer should not download other projects' tests by default is that typically people don't think about security implications when writing tests. Drupal now has a critical security issue because of behat/mink tests - see http://seclists.org/fulldisclosure/2015/Oct/42

Obviously you're going to comeback and say we're mad for having composer dependencies installed in a webroot and yes that is not recommended. A secure Drupal install would change this to be outside webroot. But Drupal targets shared hosting and we're not alone in put vendor in a web accessible directory, Joomla and i think PHPBB might do this as well.

And .gitattributes feels like a poor solution because you'll have no idea if this breaks when the dependency moves something.

**Comment by remicollet** [comment]
> So one reason why I think composer should not download other projects' tests by default
> is that typically people don't think about security implications when writing tests. 

Please note that download != install

**Comment by glensc** [comment]
i agree with point that tests should not be installed, if you want  to run test for vendor/foo/bar, you checkout foo/bar separately and run tests there.

also you can't just run tests in vendor/foo/bar, it has no vendor present. you need to run composer install for foo/bar. 

that is running foo/bar tests need "require-dev" installed of foo/bar, that are not installed (require-dev is installed only in root project)

**Comment by glen-84** [comment]
I *really* hate having to download and store tests, docs, and examples.

It makes even less sense in production.

Since we can't rely on package authors to exclude these folders, I really don't know what the solution is.

**Comment by alcohol** [comment]
`.gitattributes` is the proper solution for this.

**Comment by glensc** [comment]
yet `.gitattributes` is github centric

**Comment by alcohol** [comment]
That covers 99%. The rest can manage.

**Comment by remicollet** [comment]
Please read again https://github.com/composer/composer/issues/4438#issuecomment-142827536

.gitattributes is only a workaround.

**Comment by glen-84** [comment]
@alcohol Like I said, that doesn't help end-users of packages. The majority of the packages in my vendor folders include tests.

Also, if it is the proper solution, then explain [this](https://github.com/composer/composer/commit/bf328697e1ae7d290fb17df2e902a652ec4d9378).

**Comment by alcohol** [comment]
We all make mistakes, @Seldaek not excluded. ;-)

**Comment by jakoch** [comment]
The Composer maintainers are promoting the usage of .gitattributes (see the comments of alcohol and naderman), while for instance Symfony dropped its usage. 

@glen-84 The removal of the docu section you linked is influenced by the decision taken over at the Symfony project to drop the usage of .gitattributes.

Please keep in mind that Composer is not a "swiss army knife" build tool.
Its only one tool of a build toolchain. 
Not all features needed in a build/release process must be part of Composer.
Using .gitattributes is a fallback solution for those without a proper release process (including a build step performing the cleanup and packaging for production).

This "exclude feature" will probably never make it into Composer. 
Because: you can easily create a standalone script, add a build tasks for ant or phing or whatnot,or punch out a Composer plugin, which binds its actions to the post package install event. Maybe its already there... e.g. https://github.com/dg/composer-cleaner or https://github.com/barryvdh/composer-cleanup-plugin Contribute over there.

**Comment by glensc** [comment]
why put the burden of cleaning up to each user of library, if library maintainer can just make make simple change in composer.json metadata

not balanced option in my opinion.

speaking of boat, [composer archive](https://getcomposer.org/doc/03-cli.md#archive) is bloat as well. i can just to `composer install` and `tar cf` similarily from commandline. yet such command exists there.

as both tar nor composer archive look the actual files that came from repository, they just take any file, thus editor backups, temp files, everything!

**Comment by alcohol** [comment]
Because why should the burden of making this decision be with the maintainer? Why should he make this decision for all users?

@glensc I completely agree that `composer` already contains more bloat than it should. Unfortunately it is not up to me to decide what to do about that. I can do my best though to prevent more bloat from being added.

**Comment by glensc** [comment]
what decision? author just marks files as "tests" or "doc" (thinking like PEAR has), he knows what is what. of course those attributes should be completely optional. and library users can send PR to author, so there's no real burden.

PS: "bin" type already exists (if comparing  to PEAR)

**Comment by alcohol** [comment]
Again, why should he make that decision for all users? What if some users want to keep doc and exclude tests? What if some users want to keep tests but exclude docs? This is a user choice. This completely depends on context. And this has nothing to do with Composer.

**Comment by glen-84** [comment]
> @glen-84 The removal of the docu section you linked is influenced by the decision taken over at the Symfony project to drop the usage of .gitattributes.

I know, and it makes no sense at all. [This](https://github.com/symfony/symfony/issues/6605#issuecomment-60309257) comment (and the one above it), explain how it should be done. Composer already has `--prefer-source`, why not use it?

**Comment by glensc** [comment]
composer has commandline support, so there can be options like `--no-tests`, `--no-doc`, like there is `--no-dev`. it absolutely does not have to be default option. and it must not be default. at least not initially (`--dev` vs `--no-dev` changed their default over time for example).

**Comment by alcohol** [comment]
So you want to add more bloat? No thanks. I would much rather remove existing bloat.

**Comment by alcohol** [comment]
And seriously, this should be a part of your build process and can be solved with existing basic tools such as `rm` and `find` (I don't know the Windows equivalents but I'm sure they exist).

**Comment by Seldaek** [comment]
Just for the record, the reason I removed the mention about excluding tests from the docs is not because I think it's a bad idea, but because I think it's not composer's job, and I don't want to take part in endless debates about this. Both sides have good points, and ultimately it's a problem that does not really interest me.

I think there are possible technical solutions that would enable both sides to get what they want, so I wish someone would just work on that instead of arguing endlessly about who is right. The two points of view have valid reasons to want their thing, because they are motivated by completely different requirements. You can't possibly convince the other.
